app/ModelFunctions/LDAPFunctions.php

```diff
@@ -0,0 +1,377 @@
+			// no objects found
+			Logs::notice(__METHOD__, __LINE__, 'LDAP search returned non-array result.');
+
+			return false;
+		}
+
+		// Only accept one response
+		if ($result['count'] != 1) {
+			Logs::notice(__METHOD__, __LINE__, 'LDAP search returned ' . $result['count'] . ' results while it should return 1!');
+
+			return false;
+		}
+
+		$user_result = $result[0];
+
+		// general user info
+		$info['dn'] = $user_result['dn'];
+		$info['fullname'] = $user_result[$this->getConf('cn', 'cn')][0];
+		if (array_key_exists($this->getConf('mail', 'mail'), $user_result)) {
+			$info['email'] = $user_result[$this->getConf('mail', 'mail')][0];
+		} else {
+			$info['email'] = '';
+		}
+		// cache the info for future use
+		$this->user_info[$user] = $info;
+
+		return $info;
+	}
+
+	/**
+	 * Check user+password.
+	 *
+	 * Checks if the given user exists and the given
+	 * plaintext password is correct by trying to bind
+	 * to the LDAP server
+	 *
+	 * @param string $user
+	 * @param string $pass
+	 *
+	 * @return bool
+	 */
+	public function check_pass($user, $pass)
+	{
+		if (!$this->_openLDAP()) {
+			return false;
+		}
+
+		// indirect user bind
+		if ($this->getConf('binddn') && $this->getConf('bindpw')) {
+			// use superuser credentials
+			if (!@ldap_bind($this->con, $this->getConf('binddn'), $this->getConf('bindpw'))) {
+				Logs::notice(__METHOD__, __LINE__, 'LDAP bind as superuser failed.');
+
+				return false;
+			}
+			$this->bound = 2;
+		} elseif ($this->getConf('binddn') &&
+			$this->getConf('usertree') &&
+			$this->getConf('userfilter')
+		) {
+			// special bind string
+			$dn = $this->_makeFilter(
+				$this->getConf('binddn'),
+				['user' => $user, 'server' => $this->getConf('server')]
+			);
+		} elseif (strpos($this->getConf('usertree'), '%{user}')) {
+			// direct user bind
+			$dn = $this->_makeFilter(
+				$this->getConf('usertree'),
+				['user' => $user, 'server' => $this->getConf('server')]
+			);
+		} else {
+			// Anonymous bind
+			if (!@ldap_bind($this->con, $this->getConf('binddn'), $this->getConf('bindpw'))) {
+				Logs::notice(__METHOD__, __LINE__, 'LDAP: can not bind anonymously');
+
+				return false;
+			}
+		}
+
+		// Try to bind to with the dn if we have one.
+		if (!empty($dn)) {
+			// User/Password bind
+			if (!@ldap_bind($this->con, $dn, $pass)) {
+				Logs::notice(__METHOD__, __LINE__, "LDAP: bind with $dn failed");
+
+				return false;
+			}
+			$this->bound = 1;
+
+			return true;
+		} else {
+			// See if we can find the user
+			$info = $this->get_user_data($user);
+			if (empty($info['dn'])) {
+				return false;
+			} else {
+				$dn = $info['dn'];
+			}
+
+			// Try to bind with the dn provided
+			if (!@ldap_bind($this->con, $dn, $pass)) {
+				Logs::notice(__METHOD__, __LINE__, "LDAP: bind with $dn failed");
+
+				return false;
+			}
+			$this->bound = 1;
+
+			return true;
+		}
+	}
+
+	protected function getConf(string $key, int|bool|string|null $default = null): int|bool|string|null
```

Comment:
Is this method really necessary? I assume if you have introduced proper class constants for the configuration options, then this method can be nuked.